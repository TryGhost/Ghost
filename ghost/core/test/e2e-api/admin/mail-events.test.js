const assert = require('assert/strict');
const {MailEventService} = require('@tryghost/mail-events');
const {agentProvider, matchers, mockManager} = require('../../utils/e2e-framework');
const configUtils = require('../../utils/configUtils');
const models = require('../../../core/server/models');

const {anyContentVersion, anyEtag} = matchers;

describe('Mail Events API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();

        mockManager.mockLabsEnabled(MailEventService.LABS_KEY);
    });

    it('Can add a mail event', async function () {
        configUtils.set(MailEventService.CONFIG_KEY_PAYLOAD_SIGNING_KEY, 'foobarbaz');

        const payload = {
            // The signature is based on the previous config value as well as the
            // "mail_events" array below. If you change any of these values, you will need to
            // update the signature otherwise the request will fail
            signature: '51ab01400f9a78669733d85fcf344401f5da648f8c95707bc06da0456cb99fbc',
            mail_events: [
                {
                    id: 'Ase7i2zsRYeDXztHGENqRA',
                    timestamp: 1521243339.873676,
                    event: 'opened',
                    message: {
                        headers: {
                            'message-id': '20130503182626.18666.16540@sandboxb052085d6a7b401bb117d3a432d1d659.mailgun.org'
                        }
                    },
                    recipient: 'alice@example.com'
                }
            ]
        };

        await agent
            .post('/mail_events/', {
                headers: {
                    'content-type': 'application/json'
                }
            })
            .body(payload)
            .expectStatus(200)
            .matchBodySnapshot({})
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            });

        const storedMailEvent = await models.MailEvent.findOne({id: 'Ase7i2zsRYeDXztHGENqRA'});

        assert.ok(storedMailEvent, 'Expected mail event was not found in the database');
    });
});
