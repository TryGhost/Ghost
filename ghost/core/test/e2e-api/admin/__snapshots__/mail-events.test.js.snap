// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mail Events API Can add a mail event 1: [body] 1`] = `Object {}`;

exports[`Mail Events API Can add a mail event 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "2",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
